### PR TITLE
doc: clarify when to use `pub`

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

File: docs/style.md [hunk 1]
```diff
@@ -196,6 +196,46 @@ external crates. The surprise factor for derivations sharing a name with the sta
...
+    pub(crate) first_name: String,
+}
+
+pub(crate) mod auth {
+    pub(crate) check_password(password: &str) -> Result<(), ()>
+}
+
+// BAD
+pub(crate) struct Person {
+    pub first_name: String,
```
Comment (pompon0) [inline docs/style.md:228]: what is the difference between pub and pub(crate) within a pub(crate) struct?

Comment (mina86) [reply docs/style.md:228]: Honestly I don’t like this.  It’s just more syntax and the rationale doesn’t really apply.  I can see how when you’re in a middle of a module you don’t necessarily know whether that module is public or not, but within a struct the start of the struct is just a few lines above.  I’m more sympathetic to `impl` blocks.

**Review by nagisa — APPROVED** [review]

File: docs/style.md (hunk 1 above)
Comment (pompon0) [reply docs/style.md:228]: ok, but what if the intent is: a struct is currently pub(crate), but if it was ever to become pub, all the fields should be pub as well. In that case, having a pub field is a clear indication that whoever the user of the struct is, they should have access to the given field/method. In fact I see hardly any use of pub(crate) fields/methods - either something is an implementation detail of the type, or not.

Comment (pompon0) [reply docs/style.md:228]: > That isn’t necessarily true just like it’s not necessarily true that all methods of a struct are either private or pub.

That was an assumption of my example, not a conclusion.

File: docs/style.md [hunk 2]
```diff
@@ -196,6 +196,46 @@ external crates. The surprise factor for derivations sharing a name with the sta
...
+### Visibility
+
+"Minimize visibility" is a standard best practice, but rust has some extra
+specifics. In Rust, there are two granularities for visibility: visibility of
+items within a single crate, and visibility of items across the crates.
+Cross-crate APIs often form important architectural boundaries affecting the
+whole application, while what happens inside a crate concerns only the crate
+itself.
+
+The difference between `pub` and `pub(crate)` is big, while the difference
```
Comment (akhi3030) [inline docs/style.md:208]:
```suggestion
The difference between `pub` (making items visible to other crates) and `pub(crate)` (making items visible within the crate) is big, while the difference
```

File: docs/style.md [hunk 3]
```diff
@@ -196,6 +196,46 @@ external crates. The surprise factor for derivations sharing a name with the sta
...
+between `pub(crate)` and private is relatively small. Be mindful when exposing
+APIs across crates with `pub`. Similarly, be mindful when introducing new
+cross-crate dependencies by editing `Cargo.toml`.
```
Comment (akhi3030) [inline docs/style.md:211]: nit: Adding additional dependencies seems to be orthogonal to advise about visibility.  I think we could do with having a separate section about dependencies.

```suggestion
APIs across crates with `pub`.
```

File: docs/style.md [hunk 4]
```diff
@@ -196,6 +196,46 @@ external crates. The surprise factor for derivations sharing a name with the sta
...
+
+Use `pub` only for items reachable outside of the crate, don't rely on effective
```
Comment (akhi3030) [inline docs/style.md:213]:
```suggestion
Use `pub` only for items that must be reachable from outside of the crate, don't rely on effective
```

**Review by akhi3030 — APPROVED** [review]
Thanks!

File: docs/style.md (hunk 1 above)
Comment (mina86) [reply docs/style.md:228]: Got it.  So I guess the complaint here is that the example is poorly labelled.  It’s not bad to use ‘pub’ on fields of a ‘pub(crate)’.  It’s bad to use ‘pub’ on fields which are meant to be crate-public.

Comment (pompon0) [reply docs/style.md:228]: yes

Comment (matklad) [reply docs/style.md:228]: > It’s not bad to use ‘pub’ on fields of a ‘pub(crate)’. It’s bad to use ‘pub’ on fields which are meant to be crate-public.

So this is debatable. It is indeed my intention to declare `pub` field of a `pub(crate)` struct to be bad. And I definitely acknowledge that `pub(crate)` on fields is noisy syntactically. 

My arguments for still using `pub(crate)` even in this scenario are: 

* while it doesn't take a lot of time to look at the struct definition, it takes some attention: you need to combine info from two places
* it breaks "`pub` is always public" invariant, so you go from 0 thinking about pub to epsilon thinking, and to me the difference between 0 and eps seems significant, even if eps is relatively small
* it makes consistency harder: we'd now have to have a rule for *not* using `pub(crate)` on fields unless the struct is `pub`.
* The suggested rule here is enforceable automatically by `-Dunreachable-pub` built-in lint, and that `rustc` lint would complain about pub fields of non-pub struct. I am not entirely sure we should enable this lint (it has some false positives at the moment I think).

Comment (mina86) [reply docs/style.md:227]: I had thought about it a little more and while my opinion was at -0 it now shifted to -1 for this PR.  This is because I realised visibility is the least interesting aspect of what `pub` does.  It’s all about level of encapsulation.  When I see `pub` to me it’s signifies that the field, method or function encapsulates internal logic of the type or module and I probably can’t make things inconsistent if I use it.  On the other hand, `pub(crate)` indicates to me that I should avoid using that unless I understand invariants of given type or module.

For example, imagine `struct Complex { pub re: f64, pub im: f64 }`. Making the fields public makes sense because the type will work correctly for whatever finite values someone might put there.  On the other hand, consider `struct Complex { pub(crate) abs: f64, pub(crate) arg: f64 }` where the type has an invariant that 0 ≤ arg < τ.  Now the `pub(crate)` indicates that some abstraction layer is being bypassed if the field is used directly.  This hints that more care needs to be taken when accessing those fields directly.

In yet other words, `pub(crate)` means ‘don’t use unless you’re implementing this type/module’.  If we mandate `pub(crate)` everywhere, we loose that information for crate-private types.

Comment (matklad) [reply docs/style.md:227]: Yeah, I think I understand this argument! 

I would say that "encapsulation boundary" and "visibility" are orthogonal. There can be `pub` APIs which are non-the-less escape hatches for breaking abstraction boundaries (ie, `StoreUpdate::set_raw_bytes`). You also don't see visibility at the call-site, so using visibility to mark "this can break invariants" makes the job tough for the reader. It seems that in general all API should maintain invariants, unless the API itself is clearly marked as "this can violate invariant". 

Stated from the user's point of view, if you are using any type's API anywhere outside the implementation of the type, you shouldn't worry about API potentially breaking some invariant: if it could break an invariant, this should be very clear from the API itself.


For the `Complex` example, 

```rust
pub struct Complex { 
  pub(crate) abs: f64, 
  // Invariant:  0 ≤ arg < τ
  pub(crate) arg: f64 
}
```

I would push back against the design that exposes invariant-breaking fields at all. I would strongly prefer the following API instead: 

```rust
pub struct Complex { 
  abs: f64, 
  // Invariant:  0 ≤ arg < τ
  arg: f64 
}

impl Complex {
  pub fn abs(&self) -> f64;
  pub fn with_abs(&self, abs: f64) -> Result<Complex, InvalidAbs>;
  pub fn with_abs_unchecked(&self, abs: f64) -> Complex;

  pub fn arg(&self) -> f64;
  pub fn with_arg(&self, args: f64) -> Result<Complex, InvalidArg>;
  pub fn with_arg_unchecked(&self, args: f64) -> Complex;
}
```

The choice of visibility for `with_arg_unchecked` seems to be orthogonal to the choice of whether to provide this function at all. 

A real world example here would be [`Ratio::new_raw`](https://docs.rs/num-rational/latest/num_rational/struct.Ratio.html#method.new_raw). 

Questions: 

1. Do we need rules here at all? I'd very much like one, as "what should be visibility of a method in a non-pub type" is a frequent question with several plausible answers, and I want to normalize this.
2. If we don't like "pub = inter-crate" for fields&methods, what about the following rule: 
    > Make fields and method either `pub` or private
3. If we want to explicitly allow for invariant-breaking `pub(crate)`, I'd love to see a suggested wording.

Comment (mina86) [reply docs/style.md:227]: > The choice of visibility for with_arg_unchecked seems to be orthogonal to the choice of whether to provide this function at all.

This is only true if you exclude the possibility that we want to provide this function to the crate (so that it’s usage is contained and easier to audit) but not as a public API.  I don’t see why ‘everything is either public or private’ would be a good rule to have.

> Do we need rules here at all? I'd very much like one, as "what should be visibility of a method in a non-pub type" is a frequent question with several plausible answers, and I want to normalize this.

We don’t.  The answer to ‘what should be visibility of a method in a non-pub type’ is exactly the same as to ‘what should be visibility of a method in a pub type.’  If it’s part of public interface, make it `pub`.  If it’s not, make it module or crate public.  At this point it becomes question how to organise code (to take advantage of Rust’s visibility rules) rather than about visibility itself.

> Make fields and method either pub or private

This sounds like Java’s philosophy of always providing getters and setters.  IMO it only serves to increase boilerplate code.  Furthermore, since Rust cannot pierce through getters it may make it impossible to take mutable references to multiple fields of a struct.

> If we want to explicitly allow for invariant-breaking pub(crate), I'd love to see a suggested wording.

Invariant-breaking is a bit of a red herring here.  We want to allow `pub(crate)` and whether it’s invariant-breaking or not is separate issue.  And if field, method or function isn’t meant as public interface but code outside of enclosing module needs to access it then it needs to be `pub(crate)`.  There are myriad API combinations so I certainly can’t come up with short guideline encompassing all that.

Comment (matklad) [reply docs/style.md]: Revised the wording to not have any specific rules for fields and methods, PTAL!

File: docs/style.md [hunk 5]
```diff
@@ -196,6 +196,53 @@ external crates. The surprise factor for derivations sharing a name with the sta
...
+// BAD
+pub(crate) mod auth {
+    pub check_password(password: &str) -> Result<(), ()>
+}
+```
+
+**Rationale:** it's useful to tell at a glance that `pub` items form crate's
+public API without checking all the ancestor modules.
+
+For fields and methods, it's OK to use `pub` even if the type itself is not
```
Comment (nagisa) [inline docs/style.md:231]: I would mark this somehow that this text is just a note, clarification, rather than a part of the rationale, given the structure of the rest of the document.

Comment (matklad) [reply docs/style.md:231]: I think it's clear enough from context. Rationale is always a single paragraph, and we do use that in, eg, https://github.com/near/nearcore/blob/c1ae98f1bc1c9978ca7f6399fedc9542649558cc/docs/style.md#avoid-references-to-copy-types

**Comment by mina86** [comment]
LGTM

**Review by pompon0 — APPROVED** [review]

File: docs/style.md [hunk 6]
```diff
@@ -196,6 +196,53 @@ external crates. The surprise factor for derivations sharing a name with the sta
...
+
+```rust
+// GOOD
+pub(crate) struct Person {
+    pub first_name: String,
+    pub last_name: String,
+}
+
+impl Person {
+    pub fn full_name(&self) -> String;
```
Comment (akhi3030) [inline docs/style.md:242]: So in this example, can code outside of the crate access `full_name`?  Does that depend upon the visibility of `struct Person`?  If the visibility of `full_name` can change depending upon the visibility of `struct Person`, then doesn't the argument from above apply here?  `impl Person` can be very far away from `struct Person` and I will have to hunt down that definition to confirm what kind of visibility `full_name` has.

Comment (mina86) [reply docs/style.md:242]: I don’t understand that use-case to be honest.  If you think something shouldn’t be public, make it `pub(crate)`.  Why does it matter that a struct has a `pub` field?

Comment (akhi3030) [reply docs/style.md:242]: I am looking at it from the reader's point of view.  The rationale above says: 

```
**Rationale:** it's useful to tell at a glance that `pub` items form crate's public API without checking all the ancestor modules.
```

So the way I see it, if you have a pub method in a non-pub struct, that is breaking this rationale.

Comment (mina86) [reply docs/style.md:242]: Ah, yes, I see your point.  To be honest I don’t agree with that rationale either but I didn’t want to drag the discussion. ;)

**Comment by akhi3030** [comment]
I think this PR is now abandoned.
